homeassistant/components/media_player/yamaha.py

```diff
@@ -49,8 +57,40 @@ def update(self):
             self._pwstate = STATE_OFF
         self._muted = self._receiver.mute
         self._volume = (self._receiver.volume/100) + 1
-        self._current_source = self._receiver.input
-        self._source_list = list(self._receiver.inputs().keys())
+
+        if self.source_list is None:
+            self.build_source_list()
+
+        current_source = self._receiver.input
+        if self._mapping is not None:
+            if current_source in list(self._mapping.keys()):
+                self._current_source = self._mapping[current_source]
+            else:
+                self._current_source = current_source
+        else:
+            self._current_source = current_source
+
+    def build_source_list(self):
+        """Build the source list."""
+        self._source_list = []
+
+        if self._mapping is not None:
+            self._reverse_mapping = {}
+            mapping_list = list(self._mapping.keys())
+            for source, alias in self._mapping.items():
+                self._reverse_mapping[alias] = source
+        else:
+            mapping_list = []
+
+        source_list = list(self._receiver.inputs().keys())
+        for source in source_list:
```

Comment:
iterating over a dict in python acts exactly as though it was a list of the keys ie:

```python
$ python3
Python 3.4.3 (default, Mar 31 2016, 20:42:37) 
[GCC 5.3.1 20151207 (Red Hat 5.3.1-2)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> d = {'one': 1, 'two': 2, 'three': 3, 'four': 4, 'five': 5}
>>> for key in d:
...     print(key)
... 
four
two
five
three
one
```

You could replace this with:

    for source in self._receiver.inputs()